### ISSUE TITLE
Re-evaluate <Resources /> after logout and subsequent login as another user

### ISSUE THREAD

**Issue by rmuradov** [marmelab/aor-permissions#27]
Hi there!
Perhaps that ticket could be marked as feature request
So, details:
aor - v.1.2.0
aor-permissions - v1.1.0

here is the snippet of my app

```javascript
import {
    Admin
} from 'aor-permissions';

                <Admin
                    appLayout={AppLayout}
                    menu={AppMenu}
                    authClient={authClient}
                    restClient={restClient}
                    customReducers={{auth: authReducer}}
                    messages={translations}
                >
                    <Resource
                        name="users"
                        list={UsersList}
                        create={UserCreate}
                        edit={UserEdit}
                        remove={UserDelete}
                        permissions={'manager'}
                    />
                    <Resource
                        name="clients"
                        list={ClientsList}
                        create={ClientCreate}
                        edit={ClientEdit}
                        remove={ClientDelete}
                        permissions={['advisor', 'manager']}
                    />
                </Admin>
```

I faced the following problem:
1) I logged in as `UserA` which has role, lets say, `'manager'` based on that role I've exposed to him some of the resources - `Users` and `Clients`
2) When I logged out and logged in immediately as `UserB` which has role `'advisor'` I was still able to see `<Resource name="users"/>` in my sidebar menu despite inappropriate permissions. Moreover the resource pages were also available to me (list, create, edit pages).

Only after page refresh (Ctrl + F5) the `<Resources/>` were re-evaluated against permissions again and `<Resource name="users" />` was not avail anymore to `UserB`

In other words current implementation exposes to unprivileged users resources that were available to previously logged in user.

**Comment by djhi** [comment]
Can you share the code of your authClient ?

**Comment by rmuradov** [comment]
Sure
here it is:

```javascript
import { AUTH_LOGIN, AUTH_LOGOUT, AUTH_CHECK, AUTH_ERROR } from 'admin-on-rest';
import { API } from '../configs/index';

import { AUTH_GET_PERMISSIONS } from 'aor-permissions';

import {apiErrorsParser} from './apiErrorsParser';

export default (type, params) => {
    console.log(type);

    switch(type) {
        case AUTH_LOGIN:
            const { username, password } = params;

            const request = new Request(
                [API.protocol, API.basePass, ':', API.port, '/1.0/auth/login'].join(''),
                {
                    method: 'POST',
                    body: JSON.stringify({ email: username, password }),
                    headers: new Headers({ 'Content-Type': 'application/json' }),
                }
            );

            return fetch(request)
                .then(response => {
                    return response.text().then(text => {
                        return {
                            json: JSON.parse(text),
                            status: response.status,
                            statusText: response.statusText,
                            headers: response.headers
                        };
                    });
                })
                .then(({ status, statusText, headers, json }) => {
                    if (status < 200 || status >= 300) {
                        if (status == 500) {
                            throw new Error('Internal Server Error');
                        }
                        else {
                            throw new Error(apiErrorsParser((json && json.errors) || statusText));
                        }
                    }

                    return json;
                })
                .then((user) => {
                    localStorage.setItem('current_user', JSON.stringify(user));

                    const { api_token } = user;
                    localStorage.setItem('api_token', api_token);

                    return user;
                })
                .catch((error) => {
                    // This means that network error has occured
                    if (error.message == "Failed to fetch" || error.stack == 'TypeError: Failed to fetch') {
                        throw new Error("errors.network_error");
                    }

                    throw new Error(error);
                });

        case AUTH_LOGOUT:
            let api_token = localStorage.getItem('api_token');
            let promise = null;

            if (api_token) {
                const logout = new Request(
                    [API.protocol, API.basePass, ':', API.port, '/1.0/auth/logout'].join(''),
                    {
                        method: 'POST',
                        headers: new Headers({
                            'Content-Type': 'application/json',
                            'Authorization': `${api_token}`
                        })
                    }
                );

                localStorage.removeItem('api_token');
                localStorage.removeItem('current_user');

                promise = fetch(logout).then(response => {});
            }
            else {
                promise = Promise.resolve();
            }

            return promise;

        case AUTH_CHECK:
            return localStorage.getItem('api_token') ? Promise.resolve() : Promise.reject();

        case AUTH_GET_PERMISSIONS: {
            let current_user = localStorage.getItem('current_user');

            if (current_user) {
                current_user = JSON.parse(current_user);

                return Promise.resolve(current_user.role);
            }
            else {
                return Promise.resolve(false);
            }
        }

        case AUTH_ERROR:
            const {status} = params;

            if (status === 401) {
                localStorage.removeItem('api_token');
                localStorage.removeItem('current_user');

                return Promise.reject();
            }

            return Promise.resolve();

        case 'default': 
            return Promise.reject('Unkown method');
    }
}
```

**Comment by djhi** [comment]
I see that you are using a custom menu, right ? Can you share its code too ?

**Comment by djhi** [comment]
Nevermind, I reproduced it. Working on a fix

**Comment by djhi** [comment]
I'm afraid this won't be fixed with the permissions in their current addon form. FYI, we decided to make the permissions part of the core admin-on-rest package and I hope to ship it with the next `1.3.0` release.

**Comment by deluan** [comment]
While we wait for the 1.3.0, is there a workaround for this issue? Any way to conditionally render the resources in menu based on authorization info?

**Comment by djhi** [comment]
Well it currently works with the last version of this addon (see the README). 

The problem occurs when you sign out and sign in back.

Maybe a custom saga catching logouts and reloading the page would do the trick ? I know its ugly but I honestly don't see any other way before integrating this directly in aor

**Comment by rmuradov** [comment]
in fact custom saga is the only suitable and save place to reload page and thus re-evaluate resources. but i did this in saga that guarded against USER_LOGIN_SUCCESS event

**Comment by deluan** [comment]
@djhi Yes, sorry, I know it is working, I'm using the latest version. I meant to say "Any way to avoid the issue described here?" 

@rmuradov Do you mind sharing your code? I'm new to sagas and this is a blocker for me right now....

**Comment by rmuradov** [comment]
@deluan 

sure, there is nothing secret there and in fact it is pretty simple
the saga file:
```javascript
import { all, put, call, takeEvery } from 'redux-saga/effects';
import { push, replace } from 'react-router-redux';

import {
    USER_LOGIN,
    USER_LOGIN_LOADING,
    USER_LOGIN_SUCCESS,
    USER_LOGIN_FAILURE,
    USER_CHECK,
    USER_LOGOUT,
} from 'admin-on-rest';

const refreshApp = () => {
    // A dirty hack to overcome issue related to this ticket
    window.parent.location = '/';
}

export default function* () {
    yield all([
        takeEvery(USER_LOGIN_SUCCESS, refreshApp)
    ]);
}
```

the app file
```javascript
import loginSaga from '.somewhere';

import {
    Admin
} from 'aor-permissions';

                <Admin
                    customSagas={[loginSaga]}
                    ....
                >
```
and don't forget to add the `redux-saga` to `dependencies` and `babel-runtime` and `babel-plugin-transform-runtime` to the `devDependencies` sections of your `package.json`
otherwise you would end up with run-time errors related to generators

**Comment by deluan** [comment]
Thanks very much @rmuradov.

**Comment by djhi** [comment]
FYI, I started working on the integration into the core package: https://github.com/marmelab/admin-on-rest/pull/959

You're welcome to comment there about the implementation :)

**Comment by djhi** [comment]
It has been merged! It will be available in `1.3.0`. Hope you'll enjoy it!

**Comment by zifnab87** [comment]
hi - I have tried to incorporate the new functionality by using:

```
{
  "name": "platform-app",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "admin-on-rest": "git://github.com/marmelab/admin-on-rest.git#next",
    "react": "^15.6.1",
    "react-dom": "^15.6.1",
    "react-google-maps": "^7.1.0",
    "react-scripts": "1.0.10",
    "recharts": "^1.0.0-alpha.2",
    "string-format": "^0.5.0"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test --env=jsdom",
    "eject": "react-scripts eject"
  }
}

```
and when i start the application I get this error:

![image](https://user-images.githubusercontent.com/526062/29940436-e17e0b4e-8e97-11e7-8624-199774811ecd.png)

are there any serious breaking changes since 1.2.2?

thanks in advance!

**Comment by djhi** [comment]
Unfortunately, it's not possible to use `admin-on-rest` from the git url. We might make it possible in the future. Please open an issue on the `admin-on-rest` repository.

FYI, the `1.3.0` version will probably be released next week :)

**Comment by zifnab87** [comment]
@djhi awesome news - I will wait for the release then :)

**Comment by djhi** [comment]
**This project is discontinued. Its features have been merged in admin-on-rest 1.3.0. See [admin-on-rest documentation](https://marmelab.com/admin-on-rest/Authorization.html)**
